playbooks/continuous_delivery/ci-cd-elements.adoc

```diff
@@ -0,0 +1,252 @@
+
+The concept of immutable infrastructure can be applied to container-based solutions and deployment pipelines with the idea of progressively locking the infrastructure:
+
+image::images/ImmutableInfrastructure.png[]
+
+Pieces of the infrastructure are locked down in different phases and each infrastructure piece becomes immutable after being locked down.
+
+The* Code Build phase* locks the business logic. This is about creating a deployable artifact starting from our code. We have always done this step. +
+The* Image Build phase* locks the binary execution environment and creates a container image. After the container image is created, the business logic can be executed anywhere with no drift (for binaries and other runtime dependencies). +
+The* Topology Definition phase* locks the topology of the solution. Assuming the deployment involves a solution comprised of several components, this phase creates and locks the topology that describes which components must be deployed and their dependencies. The declarative API is leveraged to create a manifest of the compute, networking, and storage resources that are needed.
+
+Notice that we have already locked down most of the infrastructure and we haven’t deployed anything yet.
+
+The last phase performs the deployment to an environment. During deployment,  environment-dependent properties are injected, including public properties as well as confidential credentials.
+
+To comply with the principle of immutable infrastructure after deployment, the only way to introduce new change in the environment is to repeat the previous processes. This is true regardless of the nature of the change that we want to introduce, which might include:
+
+* Code release
+* Runtime updates
+* Security Patches
+* Configuration changes
+
+In traditional environments, the last three items are usually the responsibility of the operations team, and the developer team is not necessarily involved when they are performed. With immutable infrastructure and containers, deployment processes need to be radically rethought and need to involve the developer teams. These types of process change sometimes get push back.
+
+The phases of the progressing locking process that we have described can implemented in a CI/CD pipeline. The following sections discuss the types of pipelines and the types of steps that are part of a pipeline.
+
+== Types of pipelines
+
+Release management processes typically feature automated deployment procedures. An authorized user selects a software version and an environment, and then initiates the release process. 
+
+If the promotion workflow from the development environment to the production environment is a series of manually triggered deployments, this is referred to as a *point and click* pipeline.
```

Comment:
Is "Point and Click" an industry concept? First time I've heard it officially used

Reply:
I was wondering the same thing. I typically hear these referred to as manual and automatic deployments. I haven't heard them referred to as "straight through" and "point and click" pipelines.

The "straight through" description and diagram imply that the same version of the code is deployed to every environment.

It might be worth revising this section to focus on build triggers rather than types of pipelines. For example, you might want to auto-deploy to dev, but then manually promote from dev to test and from test to production.

One other cool thing about doing these builds with containers is that the _same container image_ is promoted through each step of the pipeline. You typically don't have to do a new build for each environment, which makes the application work more consistently across environments.

Reply:
It might be helpful to align the language here with the terminology that Jenkins uses: https://jenkins.io/doc/book/pipeline/#pipeline-concepts